### PR TITLE
Feature/email verification

### PR DESCRIPTION
**Pull request by emptycoder**
Add Email Verification
Fix bug with model User, when try to update
SQL Injection security
Reformat data for DB

### CODE REVIEW

File: Cooper/Controllers/AuthController.cs [hunk 1]
```diff
@@ -37,20 +37,23 @@ public IActionResult Login([FromBody]Login login)
...
 
+            login.Username = SQLInjectionSecurity(login.Username);
+            login.Password = SQLInjectionSecurity(login.Password);
+
             bool authValid = userRepository.CheckCredentials(login.Username, login.Password);
 
-            if (authValid)
-            {
-                string tokenString = new TokenFactory(login, configProvider).GetTokenString();
+            if (!authValid) return Unauthorized();
```
Comment (Hunroll) [inline Cooper/Controllers/AuthController.cs]: Multiple "returns" in one method. Codestyle.

File: Cooper/Controllers/AuthController.cs [hunk 2]
```diff
@@ -37,20 +37,23 @@ public IActionResult Login([FromBody]Login login)
...
+            //TODO: Add error: please verify your account
+            if (!userRepository.CheckVerifyByNickname(login.Username)) return Unauthorized();
```
Comment (Hunroll) [inline Cooper/Controllers/AuthController.cs]: { }, codestyle

File: Cooper/Controllers/AuthController.cs [hunk 3]
```diff
@@ -37,20 +37,23 @@ public IActionResult Login([FromBody]Login login)
...
-            {
-                return Unauthorized();
-            }
+            string tokenString = new TokenFactory(login, configProvider).GetTokenString();
+                
+            return Ok(new { Token = tokenString });
         }
 
-        
+        private string SQLInjectionSecurity(string value)
```
Comment (Hunroll) [inline Cooper/Controllers/AuthController.cs]: This method should be somewhere in DBTools or something on DB level and look like "public static string SanitizeString(string)"

File: Cooper/Controllers/AuthController.cs [hunk 4]
```diff
@@ -37,20 +37,23 @@ public IActionResult Login([FromBody]Login login)
...
+        {
+            return value.Replace("'", "").Replace("\"", "");
```
Comment (Hunroll) [inline Cooper/Controllers/AuthController.cs]: what about "--" ? Will it comment the rest of query?

File: Cooper/Controllers/RegistrationController.cs [hunk 5]
```diff
@@ -14,32 +14,70 @@ namespace Cooper.Controllers
...
         [HttpPost]
         [ProducesResponseType(200)]
         [ProducesResponseType(201)]
         public IActionResult Post([FromBody]UserRegistration user, string Password)
         {
-            bool nicknameExists = userRepository.IfNicknameExists(user.Nickname);
-            bool emailExists = userRepository.IfEmailExists(user.Email);
+            // TODO: send the proper explanation for bad-request.
+
+            if (!ModelState.IsValid) return BadRequest(ModelState);
```
Comment (Hunroll) [inline Cooper/Controllers/RegistrationController.cs]: multiple returns

File: Cooper/Controllers/RegistrationController.cs [hunk 6]
```diff
@@ -14,32 +14,70 @@ namespace Cooper.Controllers
...
+
+            user.Password = SQLInjectionSecurity(user.Password);
+            user.Nickname = SQLInjectionSecurity(user.Nickname);
+            user.Email = SQLInjectionSecurity(user.Email);
+
+            if (userRepository.IfNicknameExists(user.Nickname)) return BadRequest(ModelState);
```
Comment (Hunroll) [inline Cooper/Controllers/RegistrationController.cs]: duplicated lines. And still multiple returns.

File: Cooper/Controllers/RegistrationController.cs [hunk 7]
```diff
@@ -14,32 +14,70 @@ namespace Cooper.Controllers
...
-            if (!ModelState.IsValid || nicknameExists || emailExists)
-            {
-                return BadRequest(ModelState);
-            }
+            userRepository.Create(verify);
+            this.smtpClient.SendMail(verify.Email, "Register confirmation", "", verify.Token);
+            return Ok(userRepository.Create(user));
+        }
+
+        private string SQLInjectionSecurity(string value)
```
Comment (Hunroll) [inline Cooper/Controllers/RegistrationController.cs]: duplicated method :(

File: Cooper/Services/SmtpClientService/GmailSmtpClient.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Cooper.Services
+{
+    class GmailSmtpClient : ISmtpClient
+    {
+        public string SmtpServerName {get;} = "smtp.gmail.com";
+
+        private string from = "cooper.noreply@gmail.com";
+        private string password = "INSERT_PASSWORD_HERE";
+        private string url = "https://localhost:5001/confirm?token="; //TODO: get automatically server url
```
Comment (Hunroll) [inline Cooper/Services/SmtpClientService/GmailSmtpClient.cs]: Nooooooooo
Url is https://cooper.serve.games/confirm......

File: Cooper/Controllers/RegistrationController.cs (hunk 5 above)
Comment (emptycoder) [reply Cooper/Controllers/RegistrationController.cs]: Because, there will be different errors.

Comment (Hunroll) [reply Cooper/Controllers/RegistrationController.cs]: IActionResult result = null;
If (...)
{
result = BadRequest(...)
}
else
{
.....
}
...
...
return result;

According to codestyle, all methods should have only one "return"

File: Cooper/DAO/VerificationDAO.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                logger.Info($"Deleting token with id={id} was failed.");
+            }
+        }
+
+        public VerificationDb Get(object id)
+        {
+            VerificationDb verify = null;
+            EntityORM entity = crud.Read(id, "TOKEN", attributes, table);
+
+            if (entity != null)
```
Comment (Hunroll) [inline Cooper/DAO/VerificationDAO.cs]: { }

**Pull request merged**